File: frontend/src/types/route.ts
export interface CargoDimensions {
  height: number
  width: number
  weight: number
}

export type LocationMode = 'station' | 'coordinates' | 'live'

export interface TrainLocationInput {
  mode: LocationMode
  station_code?: string
  lat?: number
  lon?: number
}

export type SegmentPhase = 'TRAVERSED' | 'CURRENT' | 'UPCOMING'

export interface SegmentPath {
  id: string
  status: 'APPROVED' | 'HARD_BLOCKED'
  coordinates: [number, number][]
  phase?: SegmentPhase
  label?: string
}

export interface TrackSegmentDetail {
  id: string
  label: string
  phase: SegmentPhase
  distance_km: number
  progress_pct?: number
  max_height: number
  max_width: number
  max_weight: number
  congestion: number
  historical_delay_hours: number
  clearance_status: 'APPROVED' | 'HARD_BLOCKED'
  advisory?: string
}

export interface AlternateRoute {
  label: string
  reliability_score: number
  segment_ids: string[]
  estimated_hours?: number
  weather_score?: number
}

export interface TrainPosition {
  lat: number
  lon: number
  mode: LocationMode
  snapped_track?: string
  offset_km?: number
  station_code?: string
}

export interface ScoreBreakdown {
  weather: number | null
  port: number | null
  congestion: number
  historical: number
}

export interface ProvenanceSummary {
  decision_record_id: string
  traceability: { traced: number; total: number; coverage_pct: number }
  freshness: Record<string, number>
  source_modes: Record<string, number>
  decision_use: Record<string, number>
  warnings: string[]
}

export interface SourceBrief {
  id: string
  key: string
  label: string
  category: string
  authority_level: string
  official: boolean
  free_for_mvp: boolean
  requires_key: boolean
  reference_url?: string
  license_name?: string
  attribution_text?: string
  limitations: Record<string, unknown>
  freshness_policy: Record<string, number | null>
  current_status?: Record<string, unknown>
}

export interface ProvenanceRecord {
  id: string
  entity_type: string
  decision_input_role?: string
  canonical_source_type: string
  raw_source_state?: string
  source?: SourceBrief
  observed_at?: string
  fetched_at: string
  freshness: { state: string; age_seconds?: number }
  used_in_decision: boolean
  excluded_reason?: string
  transform: Record<string, string | null>
  value_summary: Record<string, unknown>
  decision_impact: Record<string, unknown>
}

export interface RouteEvidence {
  route: Record<string, unknown>
  snapshot: Record<string, unknown>
  traceability: ProvenanceSummary
  components: Array<{ role: string; label: string; status: string; record_ids: string[]; explanation: string }>
  records: ProvenanceRecord[]
  edges: Array<{ id: string; parent_record_id: string; child_record_id: string; relationship: string }>
  warnings: string[]
}

import type { EvidenceDecisionState, EvidenceKitStatus } from './status'

export interface RouteEvidenceKit {
  generated_at: string
  format_version: string
  kit_status: EvidenceKitStatus
  decision_state: EvidenceDecisionState
  reason_codes: string[]
  provider_refetch_required: false
  manifest: Record<string, unknown>
  manifest_checksum: string
  integrity: {
    verified: boolean
    record_count: number
    checksum_failures: string[]
  }
  evidence: RouteEvidence | null
  limitations: string[]
}

/**
 * Return the fail-closed reason for an EvidenceGate kit, or null when it is
 * eligible for an immediate approval/dispatch action.
 */
export function evidenceKitBlockReason(kit: RouteEvidenceKit | null | undefined): string | null {
  if (kit === undefined) return 'Evidence verification is still pending.'
  if (kit === null) return 'The route evidence kit is unavailable.'
  if (kit.decision_state !== 'READY') return `Evidence decision is ${kit.decision_state}.`
  if (kit.kit_status !== 'HOT') return `Evidence kit status is ${kit.kit_status}.`
  if (!kit.integrity.verified) return 'Evidence integrity verification failed.'
  if (kit.integrity.checksum_failures.length > 0) {
    return `${kit.integrity.checksum_failures.length} evidence checksum failure(s) were detected.`
  }
  return null
}

export interface ComplianceCheck {
  id: string
  route_id: string
  overall_status: string
  rule_pack_version: string
  disclaimer: string
  overridden: boolean
  override_reason?: string
  overridden_by?: string
  overridden_at?: string
  created_at: string
  documents: Array<{
    id: string
    document_type: string
    document_number?: string
    issuing_authority?: string
    issued_at?: string
    expires_at?: string
    checksum?: string
    verification_state: string
  }>
  overrides: Array<{ id: string; user_id: string; reason: string; created_at: string }>
  items: Array<{
    id: string
    rule_key: string
    rule_version: string
    status: string
    explanation: string
    recommended_action: string
    penalty_exposure?: string
    evidence: Record<string, unknown>
    rule_source: Record<string, unknown>
  }>
}

export interface RouteEvaluateResponse {
  route_id: string
  status: 'APPROVED' | 'HARD_BLOCKED'
  decision_state: EvidenceDecisionState
  reliability_score: number
  blocking_segment_id?: string
  estimated_hours?: number
  score_breakdown?: ScoreBreakdown
  segments: SegmentPath[]
  environmental_alerts: string[]
  provenance_summary?: ProvenanceSummary
}

export interface RouteSuggestResponse extends RouteEvaluateResponse {
  /** Stored backend route IDs for every leg in a client-composed waypoint journey. */
  route_ids?: string[]
  train_position: TrainPosition
  remaining_km: number
  eta_hours?: number
  track_details: TrackSegmentDetail[]
  alternate_routes: AlternateRoute[]
  next_station?: string
}

export interface Station {
  id: string
  name: string
  code: string
  lat: number
  lon: number
}

export interface ThreatSimulationResponse {
  original_score: number
  simulated_score: number
  degradation_pct: number
  alerts: string[]
}

import type { ConditionCategory, ConditionType } from '../maps/conditionSymbols'

export interface EnvironmentalZone {
  id: string
  type: ConditionType | 'storm' | 'dust' | 'solar' | 'fog' | 'cloud'
  coordinates: [number, number][]
}

export interface MapCondition {
  id: string
  type: ConditionType
  category: ConditionCategory
  lat: number
  lon: number
  reading?: string
  detail?: string
}

export interface OperatorLoadingWindow {
  start_time: string
  end_time: string
  manifest_reference: string
  manifest_sha256: string
}

export interface RouteHistoryItem {
  id: string
  source_station_code: string
  dest_station_code: string
  cargo_height_requested: number
  cargo_width_requested: number
  cargo_weight_requested: number
  status: 'APPROVED' | 'HARD_BLOCKED'
  dispatch_status: 'DRAFT' | 'DISPATCHED'
  reliability_score: number
  estimated_hours?: number
  dispatched_at?: string
  created_at: string
}

export interface TrainSchedule {
  id: string
  generated_route_id?: string
  train_code: string
  train_name: string
  source_station_code: string
  dest_station_code: string
  scheduled_departure: string
  scheduled_arrival: string
  berth_window_start?: string
  berth_window_end?: string
  schedule_status: 'PLANNED' | 'READY' | 'DISPATCHED' | 'CANCELLED'
  conflict_status: 'CLEAR' | 'WARNING' | 'BLOCKED'
  conflict_reason?: string
  is_demo: boolean
  dispatched_at?: string
  created_at: string
  updated_at: string
}

File: backend/app/schemas/route.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, model_validator

from app.services.congestion import CongestionSource
from app.services.port_sync import LoadingWindow, PortDataSource
from app.schemas.provenance import ProvenanceSummary


class OperatorLoadingWindow(BaseModel):
    """Vessel loading window entered by the operator from the manifest.

    Supplying this is the reliable path: there is no free public JNPT berth
    feed, so without it port alignment is excluded from the score.
    """

    start_time: datetime
    end_time: datetime
    manifest_reference: str = Field(..., min_length=3, max_length=200)
    manifest_sha256: str = Field(..., pattern=r"^[A-Fa-f0-9]{64}$")

    @model_validator(mode="after")
    def _check_order(self) -> "OperatorLoadingWindow":
        if self.end_time <= self.start_time:
            raise ValueError("end_time must be after start_time")
        return self


class _HasPortWindow(BaseModel):
    """Mixin: shared operator-window plumbing for route requests."""

    loading_window: OperatorLoadingWindow | None = None

    def operator_loading_window(self) -> LoadingWindow | None:
        if self.loading_window is None:
            return None
        return LoadingWindow(
            start=self.loading_window.start_time,
            end=self.loading_window.end_time,
            manifest_reference=self.loading_window.manifest_reference.strip(),
            manifest_sha256=self.loading_window.manifest_sha256.lower(),
        )


class CargoDimensions(BaseModel):
    height: float = Field(..., gt=0, description="Cargo height in meters")
    width: float = Field(..., gt=0, description="Cargo width in meters")
    weight: float = Field(..., gt=0, description="Cargo weight in tons")


class RouteEvaluateRequest(_HasPortWindow):
    cargo: CargoDimensions
    source_code: str = Field(..., min_length=2, max_length=10)
    dest_code: str = Field(..., min_length=2, max_length=10)
    port_id: str = Field(..., min_length=2, max_length=80)
    vessel_id: str = Field(..., min_length=1, max_length=120)
    train_arrival_hours: float = Field(
        ..., gt=0, le=720, description="Expected train arrival offset in hours"
    )


class TrainLocationInput(BaseModel):
    mode: str = Field(..., pattern="^(station|coordinates|live)$")
    station_code: str | None = None
    lat: float | None = None
    lon: float | None = None


class RouteSuggestRequest(_HasPortWindow):
    cargo: CargoDimensions
    destination_code: str = Field(..., min_length=2, max_length=10)
    location: TrainLocationInput
    port_id: str = Field(..., min_length=2, max_length=80)
    vessel_id: str = Field(..., min_length=1, max_length=120)
    train_arrival_hours: float = Field(..., gt=0, le=720)


class TrackSegmentDetail(BaseModel):
    id: UUID
    label: str
    phase: str
    distance_km: float
    progress_pct: int | None = None
    max_height: float
    max_width: float
    max_weight: float
    congestion: float
    historical_delay_hours: float
    clearance_status: str
    advisory: str | None = None


class AlternateRoute(BaseModel):
    label: str
    reliability_score: int
    segment_ids: list[UUID]
    estimated_hours: float | None = None
    weather_score: float | None = None


class TrainPosition(BaseModel):
    lat: float
    lon: float
    mode: str
    snapped_track: str | None = None
    offset_km: float | None = None
    station_code: str | None = None


class ScoreBreakdown(BaseModel):
    weather: float | None
    port: float | None
    congestion: float
    historical: float
    # Where the port figure came from, and whether it counted at all.
    port_data_source: PortDataSource = PortDataSource.UNAVAILABLE
    port_counted: bool = False
    applied_weights: dict[str, float] = {}
    # How the congestion figure was reached: seeded baseline, or baseline
    # blended with live rail/AIS telemetry.
    congestion_source: CongestionSource = CongestionSource.STATIC_ONLY
    congestion_static: float | None = None
    congestion_live_rail: float | None = None
    port_congestion_pct: float | None = None


class SegmentPathPoint(BaseModel):
    lat: float
    lon: float


class SegmentPathResponse(BaseModel):
    id: UUID
    status: str
    coordinates: list[list[float]]
    phase: str | None = None
    label: str | None = None


class RouteEvaluateResponse(BaseModel):
    route_id: UUID
    status: str
    decision_state: Literal["HARD_BLOCKED", "HOLD", "UNAVAILABLE", "READY"]
    reliability_score: int
    blocking_segment_id: UUID | None = None
    estimated_hours: float | None = None
    score_breakdown: ScoreBreakdown | None = None
    segments: list[SegmentPathResponse] = []
    environmental_alerts: list[str] = []
    provenance_summary: ProvenanceSummary | None = None


class RouteSuggestResponse(RouteEvaluateResponse):
    train_position: TrainPosition
    remaining_km: float
    eta_hours: float | None = None
    track_details: list[TrackSegmentDetail] = []
    alternate_routes: list[AlternateRoute] = []
    next_station: str | None = None


class StationResponse(BaseModel):
    id: UUID
    name: str
    code: str
    lat: float
    lon: float

    model_config = {"from_attributes": True}


class ThreatSimulationRequest(BaseModel):
    route_id: UUID
    storm_severity: float = Field(0.0, ge=0, le=100)
    solar_kp_index: int = Field(0, ge=0, le=9)
    port_congestion: float = Field(0.0, ge=0, le=100)


class ThreatSimulationResponse(BaseModel):
    original_score: int
    simulated_score: int
    degradation_pct: float
    alerts: list[str]


class RouteHistoryResponse(BaseModel):
    id: UUID
    source_station_code: str
    dest_station_code: str
    cargo_height_requested: float
    cargo_width_requested: float
    cargo_weight_requested: float
    status: str
    dispatch_status: str
    reliability_score: int
    estimated_hours: float | None = None
    dispatched_at: datetime | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class RouteDispatchResponse(BaseModel):
    route_id: UUID
    dispatch_status: str
    dispatched_at: datetime


class RouteApprovalResponse(BaseModel):
    route_id: UUID
    approval_status: Literal["APPROVED"] = "APPROVED"
    approved_by_user_id: str
    approved_by_role: str
    approved_at: datetime
    evidence_root_checksum: str = Field(pattern=r"^[0-9a-f]{64}$")


class JourneyDispatchRequest(BaseModel):
    route_ids: list[UUID] = Field(..., min_length=1, max_length=24)

    @model_validator(mode="after")
    def unique_route_ids(self) -> "JourneyDispatchRequest":
        if len(set(self.route_ids)) != len(self.route_ids):
            raise ValueError("route_ids must be unique")
        return self


class JourneyDispatchResponse(BaseModel):
    dispatch_status: Literal["DISPATCHED"] = "DISPATCHED"
    dispatched_at: datetime
    routes: list[RouteDispatchResponse]
